# Part 3: Cas Pràctic Complet - Sistema de Cotitzacions per Email
## JCM Technologies - Departament de R+D

<table style="margin: 0; text-align: left; width:100%; border: 2px solid #2e86ab;">
    <tr>
        <td style="width: 150px; vertical-align: middle; padding: 20px;">
            <img src="https://raw.githubusercontent.com/googlefonts/noto-emoji/main/png/128/emoji_u1f4e7.png" width="120" style="display: block;" />
        </td>
        <td style="padding: 20px;">
            <h2 style="color:#2e86ab; margin: 0;">Sistema Multi-Agent per Cotitzacions</h2>
            <p style="color:#333; margin-top: 10px;"><strong>Cas d'ús real:</strong> Automatitzar el procés de sol·licitud i anàlisi de cotitzacions de components electrònics via email.</p>
            <ul style="color:#333; margin-top: 10px;">
                <li>5 agents especialitzats col·laborant</li>
                <li>Parsing intel·ligent d'emails no estructurats</li>
                <li>Comparativa multi-proveïdor automatitzada</li>
                <li>Generació d'informes executius</li>
            </ul>
            <p style="color:#ff7800; margin: 10px 0 0 0;"><strong>Durada estimada:</strong> 110 minuts (~2 hores)</p>
        </td>
    </tr>
</table>

---
## 📚 Continguts

1. **Visió general del sistema** (10 min)
2. **Preparació de dades i templates** (10 min)
3. **Agent 1: BOM Analyzer** (10 min)
4. **Agent 2: RFQ Generator** (15 min)
5. **Enviar emails REALS amb Gmail API** (20 min)
6. **Dinàmica de la formació: Companys responen** (15 min)
7. **Agent 3: Email Parser** (20 min) ⭐
8. **Agents 4 i 5: Comparació i Informe** (25 min)
9. **Integració completa del sistema** (10 min)
10. **Bonus: Millores i properes passes** (10 min)

---
## 1. Visió General del Sistema

Construirem un sistema real que automatitza el procés de cotització de components electrònics.

### 1.1 El problema real

**Situació actual a moltes empreses:**

1. 📋 Tens un BOM amb 50 components
2. ✉️ Envies emails a 3-5 proveïdors demanant preus
3. 📥 Reps respostes en formats diversos:
   - Taules ben estructurades
   - Text lliure
   - PDFs
   - Formats caòtics
4. 📊 Has de comparar manualment:
   - Preus
   - Stock disponible
   - Lead times
   - Condicions de pagament
5. 📝 Generar informe per la direcció

**Temps invertit:** 2-4 hores per BOM

**Errors comuns:**
- Perdre informació en la transcripció manual
- No comparar tots els proveïdors
- No detectar condicions ocultes (MOQ, lead times llargs)
- Informes inconsistents

### 1.2 La solució: Sistema Multi-Agent

**Workflow automatitzat:**

```
INPUT: BOM Excel
  ↓
┌─────────────────────────────────────────────────┐
│ Agent 1: BOM ANALYZER                          │
│ - Llegeix Excel                                │
│ - Identifica components crítics/cars           │
│ - Prioritza per cost/criticitat                │
└─────────────────────────────────────────────────┘
  ↓ Llista components a cotitzar
┌─────────────────────────────────────────────────┐
│ Agent 2: RFQ GENERATOR                         │
│ - Genera emails professionals                  │
│ - Personalitza per proveïdor                   │
│ - Inclou detalls tècnics                       │
└─────────────────────────────────────────────────┘
  ↓ Emails RFQ generats
┌─────────────────────────────────────────────────┐
│ [ENVIAR VIA GMAIL API] → Companys de JCM      │
│ - Cada company rep un RFQ real                 │
│ - Responen simulant proveïdors:                │
│   * Grup 1: Format taula estructurada          │
│   * Grup 2: Text lliure conversacional         │
│   * Grup 3: Format caòtic/realista             │
└─────────────────────────────────────────────────┘
  ↓ Emails REALS amb cotitzacions
┌─────────────────────────────────────────────────┐
│ Agent 3: EMAIL PARSER ⭐ (clau!)               │
│ - Extreu preus de formats diversos             │
│ - Normalitza dades                             │
│ - Gestiona informació incompleta               │
│ - Detecta condicions especials                 │
└─────────────────────────────────────────────────┘
  ↓ Dades estructurades (Pydantic)
┌─────────────────────────────────────────────────┐
│ Agent 4: PRICE COMPARATOR                      │
│ - Compara preus multi-proveïdor                │
│ - Identifica millors ofertes                   │
│ - Detecta alertes (stock, lead time)           │
│ - Calcula costos totals                        │
└─────────────────────────────────────────────────┘
  ↓ Anàlisi comparativa
┌─────────────────────────────────────────────────┐
│ Agent 5: REPORT GENERATOR                      │
│ - Informe executiu                             │
│ - Taules comparatives                          │
│ - Recomanacions de compra                      │
│ - Export Markdown/JSON/Excel                   │
└─────────────────────────────────────────────────┘
  ↓
OUTPUT: Informe complet de cotització
```

**Temps invertit amb agents:** 5-10 minuts  
**Reducció:** 90-95% del temps  
**Qualitat:** Consistent, sense errors humans

### 1.3 Per què aquest cas d'ús és excel·lent?

**1. Reflecteix la realitat**
- Moltes empreses NO tenen APIs de proveïdors
- El correu electrònic és el canal principal
- Formats diversos i no estructurats

**2. Demostra el poder dels LLMs**
- Parsing de text no estructurat
- Comprensió de context
- Extracció d'informació clau
- Normalització de dades

**3. Aplicable a altres casos**
- Facturas i albarans
- Respostes a RFQs
- Comunicació amb clients
- Processament de documentació tècnica

**4. ROI immediat**
- Estalvi de temps mesurable
- Reducció d'errors
- Millor presa de decisions
- Escalable a múltiples BOMs

### 1.4 Tecnologies que usarem

**Frameworks i llibreries:**
- ✅ CrewAI (orquestració d'agents)
- ✅ Open AI (LLM)
- ✅ Pydantic (structured outputs)
- ✅ Pandas (manipulació de dades)
- ✅ Python standard library
- ✅ Gmail API (enviar i rebre emails reals)

**Dinàmica de la formació:**
- Els companys actuen com a proveïdors
- Responen amb 3 formats diferents (realistes)
- Emails 100% reals via Gmail

---
## 2. Preparació de Dades i Templates

Abans de crear els agents, preparem les dades necessàries.

### 2.0 LLM a utilitzar

En aquest cas farem servir els models d'OpenAI, en concret el gpt5-mini. 

He provat de fer-ho córrer amb els models gratuïts de Google i no me n'he sortit.

In [1]:
import os
from dotenv import load_dotenv
from crewai import Agent, Task, Crew, Process

# Carregar variables d'entorn
load_dotenv(override=True)

# Llegir variables d'entorn
#LLM_MODEL = os.getenv("GEMINI_MODEL")  
LLM_MODEL = os.getenv("OPENAI_MODEL")  

print("✅ Llibreries carregades!")
print(f"📡 Model: {LLM_MODEL}")
print(f"🔑 OpenAI API Key: ...{os.getenv('OPENAI_API_KEY')[-4:] if os.getenv('OPENAI_API_KEY') else 'NO TROBADA'}")
print(f"🔑 Google API Key: ...{os.getenv('GOOGLE_API_KEY')[-4:] if os.getenv('GOOGLE_API_KEY') else 'NO TROBADA'}")

✅ Llibreries carregades!
📡 Model: gpt-5-mini
🔑 OpenAI API Key: ...4sAA
🔑 Google API Key: ...K7g0


### 2.1 BOM Excel REAL de JCM

Usarem un BOM **real** de JCM Technologies (TESTBOM) per fer la formació més pràctica i rellevant:

In [2]:
import pandas as pd
import os
import shutil

# Crear directori si no existeix
os.makedirs('data', exist_ok=True)

# IMPORTANT: Aquest BOM és REAL de JCM Technologies
# Cal tenir el fitxer Bill_of_Materials-TESTBOM.xlsx a la carpeta del projecte

print("📋 Copiant BOM real de JCM...\n")

# Si el fitxer no està a data/, copiar-lo
source_file = 'Bill_of_Materials-TESTBOM.xlsx'
dest_file = 'data/Bill_of_Materials-TESTBOM.xlsx'

if os.path.exists(source_file) and not os.path.exists(dest_file):
    shutil.copy(source_file, dest_file)
    print(f"✅ BOM copiat a {dest_file}")
elif os.path.exists(dest_file):
    print(f"✅ BOM ja existeix a {dest_file}")
else:
    print(f"⚠️ ATENCIÓ: No s'ha trobat {source_file}")
    print(f"   Cal col·locar el fitxer a la carpeta del projecte.")

# Llegir i mostrar informació del BOM
try:
    # El BOM real té metadata a les primeres 6 files, headers a la fila 7
    df_bom = pd.read_excel(dest_file, skiprows=7)
    
    # Filtrar només files amb MATERIAL (components reals)
    df_bom = df_bom[df_bom['MATERIAL'].notna()].copy()
    # Netejar MATERIAL: eliminar espais i convertir floats a ints
    df_bom['MATERIAL'] = df_bom['MATERIAL'].astype(str).str.strip()
    # Si hi ha ".0" al final (de floats), eliminar-lo
    df_bom['MATERIAL'] = df_bom['MATERIAL'].str.replace(r'\.0+$', '', regex=True)
    # Filtrar només números
    df_bom = df_bom[df_bom['MATERIAL'].str.match(r'^\d+$', na=False)].copy()
    
    print(f"\n📊 BOM Real de JCM Technologies - TESTBOM")
    print(f"   Variant: S-TESTBOM")
    print(f"   Total components: {len(df_bom)}")
    print(f"   Total quantitat: {df_bom['Quantity'].sum():.0f} unitats\n")
    
    print("📋 Primers 10 components:")
    display(df_bom[['MATERIAL', 'Description', 'Quantity']].head(10))
    
except FileNotFoundError:
    print("\n❌ No s'ha pogut llegir el BOM.")
    print("   Assegura't que el fitxer Bill_of_Materials-TESTBOM.xlsx")
    print("   està a la carpeta del projecte o a data/")
except Exception as e:
    print(f"\n❌ Error llegint el BOM: {e}")

📋 Copiant BOM real de JCM...

✅ BOM ja existeix a data/Bill_of_Materials-TESTBOM.xlsx

📊 BOM Real de JCM Technologies - TESTBOM
   Variant: S-TESTBOM
   Total components: 23
   Total quantitat: 37 unitats

📋 Primers 10 components:


MATERIAL                               Description  Quantity
0  3003737  DIODE BRID.RECTI AM156 40V 1A D8,5*5,1MM       1.0
1  3003522    BUZZER MAG 2KHZ 1.5VDC 12MM EXTERNALTH       1.0
2  3000598     CAP CERAMIC 100N SMD 0603 16V 10% Y5V       4.0
3  3000719            CAP ALUM 10U TH 5x11MM 25V 20%       1.0
4  3003780  DIODE SCHOTTKY D BAT54A 30V 200MA SOT-23       1.0
5  3003681     DIODE ZENER BZT55C 5V6 500MW SOD80 2P       1.0
6  3003775         DIODE LED GREEN 570NM 0805 SMD SP       1.0
7  3003762              DIODE LED RED 650NM 0805 SMD       1.0
8  3003669                  DIODE 1N 4148 SMD SOD80C       3.0
9  3003494        FUSE CRYSTAL TUBE 6,3A 5X20MM FAST       1.0

### 2.2 Definir els rols de proveïdor

Durant la formació, simularem la resposta de 3 proveïdors diferents amb personalitats diverses:

In [3]:
# Configuració de proveïdors
SUPPLIERS = {
    'DigiKey': {
        'email': 'prova@digikey.com',
        'name': 'DigiKey Electronics',
        'style': 'formal_structured',  # Respostes ben estructurades
        'response_time': '4-8 hores',
        'characteristics': [
            'Taules ben formatades',
            'Informació completa',
            'Llenguatge professional',
            'Dades precises'
        ]
    },
    'Mouser': {
        'email': 'prova@mouser.es',
        'name': 'Mouser España',
        'style': 'informal_text',  # Text lliure, més casual
        'response_time': '1-2 dies',
        'characteristics': [
            'Text en paràgrafs',
            'Estil conversacional',
            'Preguntes de seguiment',
            'Castellà informal'
        ]
    },
    'Farnell': {
        'email': 'info@farnell.com',
        'name': 'Farnell',
        'style': 'chaotic_realistic',  # Format caòtic però realista
        'response_time': '2-3 dies',
        'characteristics': [
            'Format inconsistent',
            'Abreviatures',
            'Info incompleta ocasional',
            'Estil ràpid/directe'
        ]
    }
}

print("✅ Proveïdors configurats:\n")
for name, info in SUPPLIERS.items():
    print(f"📧 {name}")
    print(f"   Email: {info['email']}")
    print(f"   Estil: {info['style']}")
    print(f"   Temps resposta: {info['response_time']}")
    print()

✅ Proveïdors configurats:

📧 DigiKey
   Email: prova@digikey.com
   Estil: formal_structured
   Temps resposta: 4-8 hores

📧 Mouser
   Email: prova@mouser.es
   Estil: informal_text
   Temps resposta: 1-2 dies

📧 Farnell
   Email: info@farnell.com
   Estil: chaotic_realistic
   Temps resposta: 2-3 dies



### 2.3 Esquemes Pydantic

Definim els models de dades estructurats que usarem:

In [ ]:
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import date

class ComponentToQuote(BaseModel):
    """Component que necessita cotització"""
    part_number: str = Field(description="Codi del component")
    description: str = Field(description="Descripció tècnica")
    quantity: int = Field(description="Quantitat necessària", gt=0)
    manufacturer: str = Field(description="Fabricant")
    category: str = Field(description="Categoria")

class QuoteRequest(BaseModel):
    """Sol·licitud de cotització (RFQ)"""
    rfq_number: str = Field(description="Número de RFQ")
    company: str = Field(description="Empresa sol·licitant")
    contact_person: str = Field(description="Persona de contacte")
    contact_email: str = Field(description="Email de contacte")
    components: List[ComponentToQuote] = Field(description="Components a cotitzar")
    notes: Optional[str] = Field(default=None, description="Notes addicionals")

class ComponentQuote(BaseModel):
    """Cotització d'un component per un proveïdor"""
    part_number: str
    supplier: str
    unit_price: float = Field(description="Preu per unitat", gt=0)
    currency: str = Field(default="EUR")
    quantity_available: Optional[int] = Field(default=None, description="Stock disponible")
    lead_time: str = Field(description="Temps de lliurament")
    moq: Optional[int] = Field(default=None, description="Minimum Order Quantity")
    valid_until: Optional[str] = Field(default=None, description="Validesa del preu")
    notes: Optional[str] = Field(default=None, description="Observacions")

class SupplierResponse(BaseModel):
    """Resposta completa d'un proveïdor"""
    supplier: str
    contact_email: str
    response_date: str
    rfq_number: str
    quotes: List[ComponentQuote]
    shipping_cost: Optional[float] = Field(default=None)
    shipping_conditions: Optional[str] = Field(default=None)
    payment_terms: Optional[str] = Field(default=None)
    additional_notes: Optional[str] = Field(default=None)

class PriceComparison(BaseModel):
    """Comparativa de preus per un component"""
    part_number: str
    description: str
    quantity_needed: int
    best_quote: ComponentQuote
    all_quotes: List[ComponentQuote]
    price_range: tuple[float, float] = Field(description="(min, max)")
    total_cost_best: float = Field(description="Cost total amb millor oferta")
    savings_vs_worst: float = Field(description="Estalvi vs pitjor oferta")
    recommendation: str
    alerts: List[str] = Field(default_factory=list)

print("✅ Esquemes Pydantic definits!")
print("\nModels disponibles:")
print("  - ComponentToQuote")
print("  - QuoteRequest")
print("  - ComponentQuote")
print("  - SupplierResponse")
print("  - PriceComparison")

✅ Esquemes Pydantic definits!

Models disponibles:
  - ComponentToQuote
  - QuoteRequest
  - ComponentQuote
  - SupplierResponse
  - PriceComparison


---
## 3. Agent 1: BOM Analyzer Expert

Aquest agent llegeix el BOM i identifica quins components cal cotitzar.

### 3.1 Tool: Read Excel BOM (reutilitzem de Part 2)

Usarem la mateixa tool que vam crear a la Part 2, però millorada:

In [5]:
from crewai.tools import tool
import pandas as pd
from typing import List
import json

@tool("Read BOM Excel")
def read_bom_excel(filepath: str) -> str:
    """
    Llegeix un fitxer Excel que conté un Bill of Materials (BOM) de JCM Technologies.
    Processa el format específic de JCM amb metadata inicial i headers a la fila 7.
    Retorna una llista de ComponentToQuote en format JSON.
    
    Args:
        filepath: Ruta al fitxer Excel (ex: 'data/Bill_of_Materials-TESTBOM.xlsx')
        
    Returns:
        JSON string amb llista de ComponentToQuote objects
    """
    try:
        # Llegir Excel saltant les primeres 6 files (metadata)
        # La fila 7 (índex 6) conté els headers
        df_bom = pd.read_excel(filepath, skiprows=7)
        
        # Filtrar només files amb MATERIAL (components reals)
        df_bom = df_bom[df_bom['MATERIAL'].notna()].copy()
        df_bom['MATERIAL'] = df_bom['MATERIAL'].astype(str).str.strip()
        df_bom['MATERIAL'] = df_bom['MATERIAL'].str.replace(r'\.0+$', '', regex=True)
        df_bom = df_bom[df_bom['MATERIAL'].str.match(r'^\d+$', na=False)].copy()
    
        if len(df_bom) == 0:
            return json.dumps([])
        
        # Netejar quantitats
        df_bom['Quantity'] = pd.to_numeric(df_bom['Quantity'], errors='coerce').fillna(1).astype(int)
        
        # Categoritzar components per tipus
        def get_category(desc):
            if pd.isna(desc):
                return 'Unknown'
            desc_lower = str(desc).lower()
            if 'diode' in desc_lower:
                if 'led' in desc_lower:
                    return 'LED'
                elif 'zener' in desc_lower:
                    return 'Diode-Zener'
                elif 'schottky' in desc_lower:
                    return 'Diode-Schottky'
                else:
                    return 'Diode'
            elif 'cap' in desc_lower or 'capacitor' in desc_lower:
                if 'alum' in desc_lower or 'electrolytic' in desc_lower:
                    return 'Capacitor-Electrolytic'
                return 'Capacitor-Ceramic'
            elif 'res' in desc_lower or 'resistor' in desc_lower:
                return 'Resistor'
            elif 'ic ' in desc_lower or 'mcu' in desc_lower:
                return 'IC'
            elif 'relay' in desc_lower or 'relé' in desc_lower:
                return 'Relay'
            elif 'trans' in desc_lower:
                if 'former' in desc_lower:
                    return 'Transformer'
                return 'Transistor'
            elif 'switch' in desc_lower:
                return 'Switch'
            elif 'buzzer' in desc_lower:
                return 'Buzzer'
            elif 'fuse' in desc_lower:
                return 'Fuse/Protection'
            else:
                return 'Other'
        
        df_bom['Category'] = df_bom['Description'].apply(get_category)
        
        # Crear llista de ComponentToQuote
        components = []
        for _, row in df_bom.iterrows():
            # Obtenir manufacturer si existeix, sinó "Unknown"
            manufacturer = "Unknown"
            if 'Manufacturer' in df_bom.columns and pd.notna(row.get('Manufacturer')):
                manufacturer = str(row['Manufacturer'])
            elif 'Fabricant' in df_bom.columns and pd.notna(row.get('Fabricant')):
                manufacturer = str(row['Fabricant'])
            
            component = ComponentToQuote(
                part_number=str(row['MATERIAL']),
                description=str(row['Description']) if pd.notna(row['Description']) else "No description",
                quantity=int(row['Quantity']),
                manufacturer=manufacturer,
                category=row['Category']
            )
            components.append(component)
        
        # Retornar com a JSON string (serialitzant els objectes Pydantic)
        return json.dumps([comp.model_dump() for comp in components], indent=2, ensure_ascii=False)
        
    except FileNotFoundError:
        return json.dumps({"error": f"No s'ha trobat el fitxer {filepath}"})
    except KeyError as e:
        return json.dumps({"error": f"No s'ha trobat la columna esperada: {e}"})
    except Exception as e:
        return json.dumps({"error": f"ERROR llegint el fitxer: {str(e)}"})

print("✅ Tool 'read_bom_excel' creada (retorna ComponentToQuote en format JSON)!")

✅ Tool 'read_bom_excel' creada (processament format JCM)!


### 3.2 Tool: Provem la tool


In [16]:
# Provar la tool read_bom_excel
import json
filepath = "data/Bill_of_Materials-TESTBOM.xlsx"
result = read_bom_excel._run(filepath)

# Si és JSON vàlid, mostrar resum + primers components
try:
    data = json.loads(result)
    if isinstance(data, list):
        print(f"✅ Tool OK: {len(data)} components (ComponentToQuote)\n")
        print("Primers 3 components:")
        for i, comp in enumerate(data[:3], 1):
            desc = str(comp.get("description", ""))[:50]
            print(f"  {i}. {comp.get('part_number')} | {desc}... | qty={comp.get('quantity')} | {comp.get('category')}")
        print("\n--- JSON complet ---\n")
        print(result)
    elif isinstance(data, dict) and "error" in data:
        print(f"❌ Error: {data['error']}")
    else:
        print(result)
except json.JSONDecodeError:
    print(result)

BOM REAL DE JCM TECHNOLOGIES (TESTBOM)
Llegit de: data/Bill_of_Materials-TESTBOM.xlsx

════════════════════════════════════════════════════════════════════════════════
RESUM GENERAL
════════════════════════════════════════════════════════════════════════════════
Total de components diferents: 23
Total d'unitats necessàries: 37

Components per categoria:
  - Resistor: 3 components
  - Diode: 2 components
  - Capacitor-Ceramic: 2 components
  - LED: 2 components
  - Fuse/Protection: 2 components
  - Transistor: 2 components
  - Switch: 2 components
  - IC: 2 components
  - Buzzer: 1 components
  - Capacitor-Electrolytic: 1 components
  - Diode-Schottky: 1 components
  - Diode-Zener: 1 components
  - Relay: 1 components
  - Transformer: 1 components

════════════════════════════════════════════════════════════════════════════════
DETALL DELS COMPONENTS
════════════════════════════════════════════════════════════════════════════════

Component 1. Part Number: 3003737
   Descripció: DIODE B

### 3.2 Crear l'Agent BOM Analyzer

In [ ]:
from crewai import Agent
from dotenv import load_dotenv

load_dotenv(override=True)

agent_bom_analyzer = Agent(
    role="Expert Analitzador de BOMs",
    goal="Analitzar BOMs i identificar components que necessiten cotització prioritària",
    backstory="""
    Ets un enginyer electrònic senior amb 20 anys d'experiència a JCM Technologies.
    
    La teva especialitat és analitzar Bills of Materials i identificar components a cotitzar.
    
    Caldrà cotitzar tots els components de la Bill of Materials, excepte components 
    passius bàsics i econòmics (Resistor, Capacitor-Ceramic, Capacitor-Electrolytic).
    
    Tens accés a la tool 'Read BOM Excel', que retorna JSON amb llista de ComponentToQuote
    (part_number, description, quantity, manufacturer, category). Interpreta aquest JSON,
    filtra els passius i distribueix la resta per proveïdor.
    
    El teu anàlisi és sempre tècnic, precís i ben prioritzat.
    Proporciones llistes clares amb justificació de cada decisió.
    """,
    tools=[read_bom_excel],
    llm=LLM_MODEL
)

print("✅ Agent 'BOM Analyzer' creat!")

✅ Agent 'BOM Analyzer' creat!


### 3.3 Task per l'Agent BOM Analyzer

In [ ]:
from crewai import Task

task_analyze_bom = Task(
    description="""
    Analitza el BOM REAL de JCM del fitxer 'data/Bill_of_Materials-TESTBOM.xlsx'.
    
    1. Crida la tool 'Read BOM Excel' amb aquest fitxer. La tool retorna JSON amb una llista
       de ComponentToQuote (part_number, description, quantity, manufacturer, category).
    
    2. Interpreta el JSON: filtra els components amb category Resistor, Capacitor-Ceramic o
       Capacitor-Electrolytic (no cal cotitzar-los). La resta, DISTRIBUIR-LOS per proveïdor:
    
       **DigiKey (especialista en ICs i electrònica digital):**
       - IC
       - Tots els altres que no siguin específics d'altres proveïdors
       
       **Farnell (especialista en components de potència):**
       - Relay, Transformer, Switch
       - Tots els altres que no siguin específics d'altres proveïdors
       
       **Mouser (especialista en semiconductors discrets):**
       - LED, Diode-Zener, Diode-Schottky, Diode, Transistor
       - Tots els altres que no siguin específics d'altres proveïdors

       **No assignar a cap proveïdor:** Resistor, Capacitor-Ceramic, Capacitor-Electrolytic
    
    3. Per cada component a cotitzar, usa els part_number, description, quantity, manufacturer
       del JSON. NO inventis ni modifiquis part numbers.
       
    IMPORTANT: Els part numbers han de ser EXACTES tal com venen al JSON (ex: 3005688).
    """,
    expected_output="""
    Una llista estructurada de tots els components a cotitzar amb aquest format per cada un:
    
    ## Component [número]
    - **Part Number:** [codi exacte del BOM/JSON]
    - **Descripció:** [descripció completa]
    - **Quantitat:** [número]
    - **Fabricant:** [nom]
    - **Proveïdor(s) assignat(s):** DigiKey / Mouser / Farnell
     
    Al final, un resum:
    ## Resum
    - Total components a cotitzar: [número]
    - Per proveïdor: DigiKey X, Mouser Y, Farnell Z
    """,
    agent=agent_bom_analyzer
)

print("✅ Task 'Analyze BOM' creada!")

✅ Task 'Analyze BOM' creada!


### 3.4 Provar l'Agent BOM Analyzer

Executem aquest primer agent per veure què identifica:

In [13]:
from crewai import Crew
from datetime import datetime


# Crew temporal només amb aquest agent
crew_test_analyzer = Crew(
    agents=[agent_bom_analyzer],
    tasks=[task_analyze_bom],
    process=Process.sequential,
    verbose=True,
    max_iter=5,      
    max_rpm=60       
)

print("🚀 Executant Agent BOM Analyzer...\n")
print("=" * 70)

# Hora actual completa (HH:MM:SS)
hora_actual = datetime.now().strftime("%H:%M:%S")
print(f"Hora actual: {hora_actual}")


result_analyzer = crew_test_analyzer.kickoff()

print("\n" + "=" * 70)
print("✅ Anàlisi completada!\n")
print("📋 COMPONENTS IDENTIFICATS PER COTITZAR:\n")
print(result_analyzer)

# Hora actual completa (HH:MM:SS)
hora_actual = datetime.now().strftime("%H:%M:%S")
print(f"Hora actual: {hora_actual}")


🚀 Executant Agent BOM Analyzer...

Hora actual: 23:02:30


╭──────────────────────────────────────────── Crew Execution Started ─────────────────────────────────────────────╮
│                                                                                                                 │
│  Crew Execution Started                                                                                         │
│  Name: crew                                                                                                     │
│  ID: 09d68326-85b1-480f-8fc0-0a1c95747ada                                                                       │
│  Tool Args:                                                                                                     │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭─────────────────────────────────────────────── 🤖 Agent Started ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Expert Analitzador de BOMs                                                                              │
│                                                                                                                 │
│  Task:                                                                                                          │
│      Analitza el BOM REAL de JCM del fitxer 'data/Bill_of_Materials-TESTBOM.xlsx'.                              │
│                                                                                                                 │
│      Has de:                                                                                                    │
│      1. Llegir el fitxer Excel amb la tool disponible                                                           │
│      2. DISTRIBUIR els components segons especialització del proveïdor:                                         │
│                                                                                                                 │
│         **DigiKey (especialista en ICs i electrònica digital):**                                                │
│         - IC                                                                                                    │
│         - Tots els altres components que no siguin específics d'altres proveïdors                               │
│                                                                                                                 │
│         **Farnell (especialista en components de potència):**                                                   │
│         - Relay                                                                                                 │
│         - Transformer                                                                                           │
│         - Switch                                                                                                │
│         - Tots els altres components que no siguin específics d'altres proveïdors                               │
│                                                                                                                 │
│         **Mouser (especialista en semiconductors discrets):**                                                   │
│         - LED                                                                                                   │
│         - Diode-Zener                                                                                           │
│         - Diode-Schottky                                                                                        │
│         - Diode                                                                                                 │
│         - Transistor                                                                                            │
│         - Tots els altres components que no siguin específics d'altres proveïdors                               │
│                                                                                                                 │
│      3. Per cada component seleccionat, proporciona:                                                            │
│         - Part Number EXACTE (el codi MATERIAL com apareix al BOM)                                              │
│         - Descripció completa                                                                                   │
│         - Quantitat necessària                                                                                  │
│         - Fabricant                                                                                             │
│                                                        

Output()

Output()

╭──────────────────────────────────────────── 🔧 Agent Tool Execution ────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Expert Analitzador de BOMs                                                                              │
│                                                                                                                 │
│  Thought: Thought: you should always think about what to do                                                     │
│                                                                                                                 │
│  Using Tool: Read BOM Excel                                                                                     │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

ERROR:crewai.telemetry.telemetry:HTTPSConnectionPool(host='telemetry.crewai.com', port=4319): Read timed out. (read timeout=30)


╭───────────────────────────────────────────── ✅ Agent Final Answer ─────────────────────────────────────────────╮
│                                                                                                                 │
│  Agent: Expert Analitzador de BOMs                                                                              │
│                                                                                                                 │
│  Final Answer:                                                                                                  │
│  ## Component 1                                                                                                 │
│  - **Part Number:** 3003737                                                                                     │
│  - **Descripció:** DIODE BRID.RECTI AM156 40V 1A D8,5*5,1MM                                                     │
│  - **Quantitat:** 1 pcs                                                                                         │
│  - **Fabricant:** No especificat al BOM                                                                         │
│  - **Proveïdor assignat:** Mouser                                                                               │
│  - Justificació: Categoria "Diode" — Mouser és especialista en diodes i semiconductors discretes.               │
│                                                                                                                 │
│  ## Component 2                                                                                                 │
│  - **Part Number:** 3003522                                                                                     │
│  - **Descripció:** BUZZER MAG 2KHZ 1.5VDC 12MM EXTERNALTH                                                       │
│  - **Quantitat:** 1 pcs                                                                                         │
│  - **Fabricant:** No especificat al BOM                                                                         │
│  - **Proveïdor assignat:** DigiKey                                                                              │
│  - Justificació: Component electromechanical/general — no és específic de potència ni de semiconductors,        │
│  DigiKey cobreix components generals i ICs.                                                                     │
│                                                                                                                 │
│  ## Component 3                                                                                                 │
│  - **Part Number:** 3000598                                                                                     │
│  - **Descripció:** CAP CERAMIC 100N SMD 0603 16V 10% Y5V                                                        │
│  - **Quantitat:** 4 pcs                                                                                         │
│  - **Fabricant:** No especificat al BOM                                                                         │
│  - **Proveïdor assignat:** DigiKey                                                                              │
│  - Justificació: Condensador ceràmic (passiu) — assignat a DigiKey per a components generals/passius.           │
│                                                                                                                 │
│  ## Component 4                                                                                                 │
│  - **Part Number:** 3000719                                                                                     │
│  - **Descripció:** CAP ALUM 10U TH 5x11MM 25V 20%                                                               │
│  - **Quantitat:** 1 pcs                                                                                         │
│  - **Fabricant:** No especificat al BOM                

Output()


✅ Anàlisi completada!

📋 COMPONENTS IDENTIFICATS PER COTITZAR:

## Component 1
- **Part Number:** 3003737
- **Descripció:** DIODE BRID.RECTI AM156 40V 1A D8,5*5,1MM
- **Quantitat:** 1 pcs
- **Fabricant:** No especificat al BOM
- **Proveïdor assignat:** Mouser
- Justificació: Categoria "Diode" — Mouser és especialista en diodes i semiconductors discretes.

## Component 2
- **Part Number:** 3003522
- **Descripció:** BUZZER MAG 2KHZ 1.5VDC 12MM EXTERNALTH
- **Quantitat:** 1 pcs
- **Fabricant:** No especificat al BOM
- **Proveïdor assignat:** DigiKey
- Justificació: Component electromechanical/general — no és específic de potència ni de semiconductors, DigiKey cobreix components generals i ICs.

## Component 3
- **Part Number:** 3000598
- **Descripció:** CAP CERAMIC 100N SMD 0603 16V 10% Y5V
- **Quantitat:** 4 pcs
- **Fabricant:** No especificat al BOM
- **Proveïdor assignat:** DigiKey
- Justificació: Condensador ceràmic (passiu) — assignat a DigiKey per a components generals/passius.

##

╭──────────────────────────────────────────────── Tracing Status ─────────────────────────────────────────────────╮
│                                                                                                                 │
│  Info: Tracing is disabled.                                                                                     │
│                                                                                                                 │
│  To enable tracing, do any one of these:                                                                        │
│  • Set tracing=True in your Crew/Flow code                                                                      │
│  • Set CREWAI_TRACING_ENABLED=true in your project's .env file                                                  │
│  • Run: crewai traces enable                                                                                    │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────── Task Completion ────────────────────────────────────────────────╮
│                                                                                                                 │
│  Task Completed                                                                                                 │
│  Name: b327dec8-b9ce-48fd-89fb-1b20309c3f7f                                                                     │
│  Agent: Expert Analitzador de BOMs                                                                              │
│  Tool Args:                                                                                                     │
│                                                                                                                 │
│                                                                                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯


### 💡 Observa el resultat

L'agent hauria d'haver identificat components com:

**✅ PRIORITAT HIGH (crítics i cars):**
- **3005688** - IC MCU PIC16F18855 (microcontrolador, el "cervell" del sistema)
- **3000836** - IC regulador L7805ABD (estabilitat alimentació crítica)
- **3003026** - Relay DPDT 12V (x2 unitats, control de potència)
- **3003060** - Transformer 230VAC→12VAC (alimentació principal)

**✅ PRIORITAT MEDIUM (importants):**
- **3003807** - Transistor NPN BC847B (x4 unitats, quantitat elevada)
- **3003808** - Transistor PNP BC857B/T1
- **3003413/3003412** - Switches tàctils (interfície usuari)
- **3003522** - Buzzer magnètic (senyalització)
- **3003780** - Díode Schottky BAT54A (velocitat crítica)
- **3003681** - Díode Zener BZT55C (protecció voltatge)
- **3003737** - Díode rectificador pont AM156 (alimentació)

**✅ PRIORITAT LOW (funcionals però econòmics):**
- **3003775/3003762** - LEDs indicadors (verd/vermell)
- **3003669** - Díodes 1N4148 (x3, ús general)
- **3003494** - Fusible cristall 6.3A
- **3003506/3003504** - Portafusibles

**❌ NO inclosos (massa econòmics):**
- Resistències SMD (3003913, 3003925, 3000151) 
- Condensadors ceràmics petits (3000598)

**Aquest output serà l'input del següent agent!**

**Aquest output serà l'input del següent agent!**

---
## 4. Agent 2: RFQ Generator

Aquest agent genera emails professionals (RFQ - Request for Quotation) per enviar als proveïdors.

### 4.1 Què és un RFQ?

**RFQ (Request for Quotation)** és un document formal que s'envia als proveïdors sol·licitant preus.

**Components d'un bon RFQ:**
- ✅ Dades de l'empresa
- ✅ Persona de contacte
- ✅ Llista detallada de components (Part Number, descripció, quantitat)
- ✅ Informació tècnica rellevant
- ✅ Data límit de resposta (opcional)
- ✅ Condicions de pagament preferides (opcional)
- ✅ To professional però amigable

### 4.2 Tool: Format RFQ Email

Aquesta tool ajuda a formatar l'email de manera professional:

In [ ]:
@tool("Format RFQ Email")
def format_rfq_email(
    supplier_name: str,
    supplier_email: str,
    components_list: str,
    rfq_number: str
) -> str:
    """
    Formata un email RFQ professional per enviar a un proveïdor.
    
    Args:
        supplier_name: Nom del proveïdor (ex: 'DigiKey')
        supplier_email: Email del proveïdor
        components_list: Llista de components en format text
        rfq_number: Número de referència del RFQ
        
    Returns:
        Email formatat professionalment
    """
    from datetime import datetime
    
    today = datetime.now().strftime("%d/%m/%Y")
    
    email_template = f"""
══════════════════════════════════════════════════════════════════════
EMAIL RFQ - {rfq_number}
══════════════════════════════════════════════════════════════════════

De: compras@jcm-tech.com
Per a: {supplier_email}
Assumpte: Sol·licitud de cotització #{rfq_number} - Components electrònics
Data: {today}

──────────────────────────────────────────────────────────────────────

Benvolgut/da equip de {supplier_name},

Des de JCM Technologies, SAU, volem sol·licitar-vos una cotització per 
als següents components electrònics per a un projecte de control de portes 
automàtiques.

COMPONENTS SOL·LICITATS:
──────────────────────────────────────────────────────────────────────

{components_list}

──────────────────────────────────────────────────────────────────────

INFORMACIÓ ADDICIONAL:

- Referència RFQ: {rfq_number}
- Data límit resposta: {(datetime.now() + pd.Timedelta(days=7)).strftime("%d/%m/%Y")}
- Lliurament preferit: Vic, Barcelona, Espanya
- Condicions de pagament: 30 dies data factura

Us agrairíem que ens proporcionéssiu:
✓ Preu unitari per component
✓ Disponibilitat de stock actual
✓ Temps de lliurament estimat
✓ MOQ (Minimum Order Quantity) si aplica
✓ Validesa de la cotització

Qualsevol dubte sobre especificacions tècniques, no dubteu en contactar-nos.

Atentament,

Departament de Compres
JCM Technologies, SAU
Tel: +34 938 03 XX XX
Email: compras@jcm-tech.com
Web: www.jcm-tech.com

══════════════════════════════════════════════════════════════════════
"""
    
    return email_template

print("✅ Tool 'format_rfq_email' creada!")

### 4.3 Crear l'Agent RFQ Generator

In [ ]:
agent_rfq_generator = Agent(
    role="Especialista en Generació de RFQs",
    goal="Generar emails RFQ professionals i complets per sol·licitar cotitzacions a proveïdors",
    backstory="""
    Ets un expert en generar sol·licituds de cotització (RFQs) per components electrònics.
    
    REGLA D'OR: SEMPRE uses els part numbers EXACTES que reps del context.
    MAI inventes part numbers. MAI modifiques part numbers.
    
    Si un component té el part number 3003413, uses 3003413.
    Si un component té el part number 3003522, uses 3003522.
    
    La teva tasca és distribuir els components analitzats entre els 3 proveïdors
    segons les seves especialitats, mantenint SEMPRE els part numbers originals.
    
    Tens accés a una tool per formatar emails RFQ professionals.
    """,
    tools=[format_rfq_email],
    llm=LLM_MODEL,
    max_iter=5
)

print("✅ Agent 'RFQ Generator' creat!")

### 4.4 Task per l'Agent RFQ Generator

In [ ]:
task_generate_rfqs = Task(
    description="""
    Genera 3 RFQs (DigiKey, Mouser, Farnell) basant-te ESTRICTAMENT en l'anàlisi del BOM.
    
    CRÍTIC - USA NOMÉS PART NUMBERS DEL CONTEXT:
    - NO inventis part numbers
    - NO modifiquis part numbers
    - USA EXACTAMENT els part numbers que l'agent anterior ha analitzat
    
    Del context (task anterior), tens la llista de components prioritaris amb:
    - Part Number EXACTE
    - Descripció
    - Quantitat
    
    Distribueix aquests components entre els 3 proveïdors segons l'especialitat:
    
    DIGIKEY (ICs i digital):
    - Tots els ICs (MCU, reguladors)
    - Switches
    - Buzzer
    - Fusibles
    
    MOUSER (Semiconductors discrets):
    - Tots els díodes (Schottky, Zener, LEDs, rectificadors)
    - Tots els transistors
    
    FARNELL (Potència):
    - Relays
    - Transformers
    
    PER CADA PROVEÏDOR, genera un email RFQ professional.
    
    IMPORTANT: Cada component ha d'aparèixer amb el seu Part Number EXACTE del BOM.
    """,
    expected_output="""
    3 RFQs professionals, un per cada proveïdor.
    
    Cada RFQ ha de contenir:
    - Destinatari (email del proveïdor)
    - Assumpte
    - Cos de l'email amb:
      * Components amb Part Number EXACTE del BOM
      * Descripció
      * Quantitat
      * Informació addicional (RFQ-2025-001, data límit, etc.)
    
    Format:
    ## RFQ per DigiKey
    [contingut email complet]
    
    ## RFQ per Mouser
    [contingut email complet]
    
    ## RFQ per Farnell
    [contingut email complet]
    """,
    agent=agent_rfq_generator,
    context=[task_analyze_bom]  # ← Rep l'output de l'agent anterior
)

print("✅ Task 'Generate RFQs' creada amb distribució intel·ligent per proveïdor!")
print("\n💡 Distribució de components per especialitat:")
print("   - DigiKey: ICs i microcontroladors + components comuns")
print("   - Mouser: Díodes i transistors + components comuns")
print("   - Farnell: Relés i transformers + components comuns")
print("\n   Components comuns (switches, buzzer, fusibles): TOTS els proveïdors")

---
## 5. Enviar Emails REALS amb Gmail API

Ara ve la part emocionant: **enviarem emails REALS** amb els RFQs generats!

### 5.1 Tool: Authenticate Gmail

Primer creem una tool per autenticar amb Gmail:

In [ ]:
import os
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build

# Permisos necessaris
SCOPES = [
    'https://www.googleapis.com/auth/gmail.send',      # Enviar emails
    'https://www.googleapis.com/auth/gmail.readonly',  # Llegir emails
    'https://www.googleapis.com/auth/gmail.modify'     # Marcar com llegit
]

def authenticate_gmail():
    """
    Autentica amb Gmail API i retorna el servei.
    
    Returns:
        Gmail API service object
    """
    creds = None
    
    # Si ja existeix token.json, el carrega
    if os.path.exists('token.json'):
        creds = Credentials.from_authorized_user_file('token.json', SCOPES)
    
    # Si no hi ha credencials vàlides, autentica
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            print("🔄 Renovant token...")
            creds.refresh(Request())
        else:
            if not os.path.exists('credentials.json'):
                raise FileNotFoundError(
                    "❌ No s'ha trobat credentials.json!\n"
                    "Segueix les instruccions de la secció 5.1 per crear-lo."
                )
            
            print("🔐 Primera autenticació...")
            print("   S'obrirà un navegador per autenticar-te.")
            flow = InstalledAppFlow.from_client_secrets_file(
                'credentials.json', SCOPES
            )
            creds = flow.run_local_server(port=0)
        
        # Guardar credencials per següents vegades
        with open('token.json', 'w') as token:
            token.write(creds.to_json())
        print("✅ Token guardat a token.json")
    
    # Crear servei Gmail
    service = build('gmail', 'v1', credentials=creds)
    
    # Obtenir info del compte
    profile = service.users().getProfile(userId='me').execute()
    print(f"✅ Autenticat com: {profile['emailAddress']}")
    
    return service

print("✅ Funció authenticate_gmail() definida")

### 5.2 Provar autenticació

Executem per autenticar-nos i verificar que funciona:

In [ ]:
# Autenticar (primera vegada obrirà navegador)
try:
    gmail_service = authenticate_gmail()
    print("\n✅ Gmail API configurada correctament!")
    print("Ja pots enviar i rebre emails programàticament.")
except FileNotFoundError as e:
    print(f"\n❌ Error: {e}")
    print("\n📋 Accions necessàries:")
    print("1. Segueix les instruccions de la secció 5.1")
    print("2. Descarrega credentials.json")
    print("3. Col·loca'l a la carpeta del projecte")
    print("4. Torna a executar aquesta cel·la")
except Exception as e:
    print(f"\n❌ Error inesperat: {e}")

### 5.3 Tool: Send Email

Ara creem la tool per enviar emails:

In [ ]:
from email.mime.text import MIMEText
import base64
from crewai.tools import tool

@tool("Send Email via Gmail")
def send_email_gmail(
    to_email: str,
    subject: str,
    body: str
) -> str:
    """
    Envia un email via Gmail API.
    
    Args:
        to_email: Email destinatari (ex: 'colleague@jcm-tech.com')
        subject: Assumpte de l'email
        body: Cos del missatge (text pla)
        
    Returns:
        Confirmació de l'enviament amb ID del missatge
    """
    try:
        # Usar el servei global ja autenticat
        global gmail_service

        # Crear missatge
        message = MIMEText(body)
        message['to'] = to_email
        message['subject'] = subject
        
        # Codificar en base64
        raw_message = base64.urlsafe_b64encode(message.as_bytes()).decode('utf-8')
        
        # Enviar
        sent_message = service.users().messages().send(
            userId='me',
            body={'raw': raw_message}
        ).execute()
        
        return f"""✅ Email enviat a {to_email}"""
        
    except Exception as e:
        return f"❌ ERROR enviant email: {str(e)}"

print("✅ Tool 'send_email_gmail' creada!")

### 5.4 Tool: Fetch Emails

Tool per llegir emails rebuts:

In [ ]:
@tool("Fetch Emails from Gmail")
def fetch_emails_gmail(
    query: str = "is:unread",
    max_results: int = 10
) -> str:
    """
    Llegeix emails de la inbox de Gmail.
    
    Args:
        query: Cerca Gmail (ex: 'is:unread subject:RFQ', 'from:colleague@company.com')
        max_results: Màxim d'emails a retornar (default: 10)
        
    Returns:
        Llista d'emails amb: From, Subject, Date, Body
    """
    try:
        # Usar el servei global ja autenticat
        global gmail_service

        # Cercar missatges
        results = service.users().messages().list(
            userId='me',
            q=query,
            maxResults=max_results
        ).execute()
        
        messages = results.get('messages', [])
        
        if not messages:
            return f"ℹ️ No s'han trobat emails que coincideixin amb: '{query}'"
        
        output = f"📧 S'han trobat {len(messages)} email(s):\n\n"
        output += "=" * 80 + "\n\n"
        
        for idx, msg in enumerate(messages, 1):
            # Obtenir detalls del missatge
            msg_data = service.users().messages().get(
                userId='me',
                id=msg['id'],
                format='full'
            ).execute()
            
            # Extreure headers
            headers = msg_data['payload']['headers']
            subject = next((h['value'] for h in headers if h['name'] == 'Subject'), 'Sin asunto')
            from_email = next((h['value'] for h in headers if h['name'] == 'From'), 'Desconocido')
            date = next((h['value'] for h in headers if h['name'] == 'Date'), 'Fecha desconocida')
            
            # Extreure cos del missatge
            body = ""
            if 'parts' in msg_data['payload']:
                for part in msg_data['payload']['parts']:
                    if part['mimeType'] == 'text/plain':
                        if 'data' in part['body']:
                            body = base64.urlsafe_b64decode(
                                part['body']['data']
                            ).decode('utf-8')
                            break
            else:
                if 'data' in msg_data['payload']['body']:
                    body = base64.urlsafe_b64decode(
                        msg_data['payload']['body']['data']
                    ).decode('utf-8')
            
            output += f"EMAIL {idx}\n"
            output += "-" * 80 + "\n"
            output += f"De: {from_email}\n"
            output += f"Assumpte: {subject}\n"
            output += f"Data: {date}\n"
            output += f"ID: {msg['id']}\n"
            output += "-" * 80 + "\n"
            output += f"COS DEL MISSATGE:\n\n{body[:1000]}"  # Limitar a 1000 caràcters
            if len(body) > 1000:
                output += "\n\n[... missatge truncat ...]\n"
            output += "\n" + "=" * 80 + "\n\n"
        
        return output
        
    except Exception as e:
        return f"❌ ERROR llegint emails: {str(e)}"

print("✅ Tool 'fetch_emails_gmail' creada!")

### 5.5 Provar les tools manualment

Abans de donar-les als agents, provem que funcionen:

In [ ]:
# OPCIONAL: Prova d'enviament (canvia l'email!)
# Descomenta i executa si vols provar

#test_result = send_email_gmail._run(
#     to_email="jberingues@jcm-tech.com",  # ← Canvia això!
#     subject="Test CrewAI Gmail Integration",
#     body="Hola! Aquest és un email de prova enviat des de CrewAI.\n\nFunciona! 🎉"
# )
#print(test_result)

print("\n💡 Per provar l'enviament, descomenta el codi de dalt i executa.")
print("   Recorda canviar 'teu_email@gmail.com' pel teu email real!")

In [ ]:
# Provar lectura d'emails
print("📬 Llegint emails recents...\n")

emails = fetch_emails_gmail._run(
    query="newer_than:1d",  # Emails de les últimes 24h
    max_results=3
)

print(emails)

---
## 6. Dinàmica de la Formació: Enviar i Respondre

Ara ve la part interactiva! Cada assistent enviarà i rebrà un RFQ.

### 6.1 Dinàmica amb 3 Voluntaris

**🎭 Dinàmica de la formació:**

Com que el sistema complet s'executarà amb **OpenAI** (per evitar problemes de quota), farem una **demostració interactiva** amb participació de la gent de la formació:

```
📊 WORKFLOW:

1. Jo (formador) executo el sistema fins generar els 3 RFQs
2. El sistema ENVIA els RFQs per Gmail als 3 voluntaris:
   - Voluntari 1 → Rep RFQ de DigiKey
   - Voluntari 2 → Rep RFQ de Mouser  
   - Voluntari 3 → Rep RFQ de Farnell
3. Els 3 voluntaris responen (10-15 min)
4. Jo executo la resta del sistema que:
   - Llegeix les 3 respostes
   - Les parseja automàticament
   - Compara preus
   - Genera informe executiu
5. Tothom veu l'informe final generat
```

**📋 Rols dels voluntaris:**

```
VOLUNTARI 1: DigiKey
  - Rebràs un RFQ amb components d'ICs i digital
  - Respon amb format: Taula estructurada (professional)
  - Estil: Formal i organitzat

VOLUNTARI 2: Mouser
  - Rebràs un RFQ amb díodes i transistors
  - Respon amb format: Text lliure conversacional
  - Estil: Friendly i informal

VOLUNTARI 3: Farnell
  - Rebràs un RFQ amb relés i transformers
  - Respon amb format: Caòtic/realista
  - Estil: Desorganitzat però amb info útil
```

**⏱️ Timing:**
- Enviar RFQs: 2 min
- Voluntaris responen: 10-15 min
- Processar respostes + informe: 3-5 min
- **Total:** ~20 min

### 6.2 Templates de resposta per als 3 voluntaris

**💡 Instruccions per als voluntaris:**

Quan rebeu l'email RFQ, **utilitzeu el template corresponent al vostre rol** com a guia, però:
- ✅ Podeu canviar preus lleugerament (±10-20%)
- ✅ Podeu afegir o treure alguna línia
- ✅ Podeu personalitzar l'estil

**L'objectiu:** Que cada resposta sigui realista i lleugerament diferent.

In [ ]:
# ═══════════════════════════════════════════════════════════════════════════
# TEMPLATES DE RESPOSTA AMB COMPONENTS REALS DEL BOM JCM
# ═══════════════════════════════════════════════════════════════════════════

TEMPLATE_DIGIKEY = """
De: sales@digikey.com
Assumpte: RE: Sol·licitud de cotització #RFQ-2025-001

Benvolgut/da,

Gràcies per la vostra consulta. Us adjunto la cotització dels components sol·licitats:

Part Number | Description                              | Qty | Unit Price | Stock    | Lead Time
------------|------------------------------------------|-----|------------|----------|----------
3005688     | IC MCU PIC16F18855T-I/SS 8BIT 8KB FLASH | 1   | €2.85      | 1000+ pcs| 2 dies
3000836     | IC VOLTAGE REG LDO 5V L7805ABD SMD SO8  | 1   | €0.45      | 500+ pcs | 1 setmana
3003413     | TACT SWITCH SPST-NO 6X6 LOW TH          | 1   | €0.18      | 2000+ pcs| 3 dies
3003412     | TACT SWITCH SPST-NO 6X6MM RAISED RED TH | 1   | €0.22      | 1500+ pcs| 3 dies
3003522     | BUZZER MAG 2KHZ 1.5VDC 12MM EXTERNALTH  | 1   | €0.95      | 300+ pcs | 5 dies
3003494     | FUSE CRYSTAL TUBE 6,3A 5X20MM FAST      | 1   | €0.12      | 5000+ pcs| 2 dies

Condicions comercials:
- Preus vàlids fins: 15/02/2026
- MOQ: 10 unitats per component
- Enviament: €25 (gratuït per comandes >€500)
- Termini de pagament: 30 dies net
- Descompte per volum disponible (consultar)

Salutacions cordials,
DigiKey Sales Team
sales@digikey.com
"""

TEMPLATE_MOUSER = """
De: ventas@mouser.es
Assumpte: Cotización - Componentes electrónicos JCM Technologies

Hola,

He revisado tu solicitud de cotización y te comento lo que tenemos disponible:

Para los díodes que necesitas, tenemos bastante stock. El díode Schottky BAT54A 
(ref 3003780) está a 0,15€ por unidad, tenemos unas 2000 unidades en stock 
y te lo podemos enviar en 2-3 días. Los díodes Zener BZT55C 5V6 (3003681) 
van a 0,08€ cada uno, también con disponibilidad inmediata.

Los LEDs verde y rojo (3003775 y 3003762) están a buen precio: 0,12€ el verde 
y 0,10€ el rojo. Tenemos bastante stock de ambos, unas 5000 unidades de cada.

Para los transistores BC847B (ref 3003807) que pides 4 unidades, el precio 
es de 0,06€/ud. Los BC857B (3003808) van a 0,07€. Stock suficiente de ambos.

El díode rectificador de puente AM156 (3003737) lo tenemos a 0,35€ y los 
díodes 1N4148 (3003669) a 0,04€ cada uno.

Los plazos de entrega son normalmente de 3-5 días laborables para península.
Portes: 15€ (gratis si superas 200€ de pedido).

¿Necesitas algo más? Cualquier duda me comentas.

Saludos,
Carlos Martínez
Mouser España
ventas@mouser.es
"""

TEMPLATE_FARNELL = """
De: info@farnell.com
Assumpte: Re: Presupuesto componentes

Buenos días,

RELAY DPDT 12V ref 3003026: 2,85€/ud tenemos 180ud STOCK OK!!! 
Entrega 24-48h península

TRANSFORMER 230VAC→12VAC 2.8VA (3003060) -> precio 4.50 EUR 
stock: 45 pcs, lead time 1 semana aprox

Switches tactiles:
- 3003413 (6x6 LOW): 0.19€ 
- 3003412 (6x6 RAISED RED): 0.24€
Stock ambos OK (2000+ uds)

Buzzer 3003522: 0,98€ tenemos 250ud entrega inmediata

Fusible 6,3A ref 3003494 -> 0,11 EUR stock SI!!! 8000pcs
Portafusibles 3003506: 0,35€ (3000 uds disponibles)

Todos los precios SIN IVA
Portes: 18€ (pedidos >350€ portes gratis)
Forma pago: transferencia 30 días o tarjeta

Saludos
Dpto. Ventas Farnell
info@farnell.com
"""

print("=" * 80)
print("📋 TEMPLATES DE RESPOSTA AMB COMPONENTS REALS JCM")
print("=" * 80)
print()
print("✅ 3 templates creats amb components del BOM TESTBOM:")
print()
print("1️⃣  TEMPLATE_DIGIKEY")
print("   • Format: Taula estructurada (professional)")
print("   • Components: ICs, Switches, Buzzer, Fuse")
print("   • Estil: Formal i organitzat")
print()
print("2️⃣  TEMPLATE_MOUSER")
print("   • Format: Text lliure conversacional (castellà)")
print("   • Components: Díodes (tots els tipus) i Transistors")
print("   • Estil: Friendly i informal")
print()
print("3️⃣  TEMPLATE_FARNELL")
print("   • Format: Caòtic/realista (mix català/castellà)")
print("   • Components: Relays, Transformer, Switches, Buzzer, Fuse")
print("   • Estil: Desorganitzat però amb info útil")
print()
print("=" * 80)
print("💡 NOTA IMPORTANT per als assistents:")
print("=" * 80)
print()
print("Aquests templates són EXEMPLES per ajudar-vos a respondre.")
print("Podeu (i hauríeu de):")
print("  • Canviar els preus lleugerament")
print("  • Afegir o treure alguna línia")
print("  • Modificar l'estil segons el vostre proveïdor")
print()
print("L'objectiu és que cada resposta sigui LLEUGERAMENT DIFERENT")
print("per simular la varietat real de respostes de proveïdors.")
print()
print("=" * 80)

### 6.3 Agent Email Sender (actualitzat amb Gmail)

Actualitzem l'agent RFQ Generator per usar la nova tool de Gmail:

In [ ]:
from crewai import Agent

agent_email_sender = Agent(
    role="Email Sender",
    goal="Send 3 RFQ emails using the Gmail tool",
    backstory="You send emails via Gmail API. Use the tool for each email.",
    tools=[send_email_gmail],
    llm=LLM_MODEL,
    reasoning_effort="none",
    allow_delegation=False  # ← AFEGIR: No delegar a altres agents
)

print("✅ Agent 'Email Sender' creat amb Gmail tool!")

### 6.4 Configuració dels emails dels 3 voluntaris

**⚠️ FORMADOR:** Abans d'executar, configura els emails dels 3 voluntaris que participaran:

In [ ]:
# ═══════════════════════════════════════════════════════════
# CONFIGURACIÓ DELS 3 VOLUNTARIS
# ═══════════════════════════════════════════════════════════
# Formador: Omple amb els emails dels 3 voluntaris escollits

# El MEU email (formador)
MY_EMAIL = "jberings01@gmail.com"  # ← Email del formador

# Emails dels 3 voluntaris
VOLUNTARI_DIGIKEY = "jberingues@jcm-tech.com"   # ← Email Voluntari 1 (DigiKey)
VOLUNTARI_MOUSER = "jberingues@jcm-tech.com"    # ← Email Voluntari 2 (Mouser)
VOLUNTARI_FARNELL = "jberingues@jcm-tech.com"   # ← Email Voluntari 3 (Farnell)

# ═══════════════════════════════════════════════════════════

print("✅ Configuració dels voluntaris:")
print(f"   Formador: {MY_EMAIL}")
print(f"\n   Voluntari DigiKey:  {VOLUNTARI_DIGIKEY}")
print(f"   Voluntari Mouser:   {VOLUNTARI_MOUSER}")
print(f"   Voluntari Farnell:  {VOLUNTARI_FARNELL}")
print(f"\n💡 Quan s'executin les tasks, s'enviarà un RFQ a cadascun d'ells.")

### 6.5 Tasks: Enviar RFQs als 3 voluntaris

Crearem 3 tasks separades per enviar cada RFQ al voluntari corresponent:

In [ ]:
task_send_all_rfqs = Task(
    description=f"""
    Executa aquestes 3 accions en ordre:
    
    1. Usa 'Send Email via Gmail' amb:
       - to_email: {VOLUNTARI_DIGIKEY}
       - subject: "RFQ DigiKey - Sol·licitud de cotització #RFQ-2025-001"
       - body: [RFQ DigiKey del context]
    
    2. Usa 'Send Email via Gmail' amb:
       - to_email: {VOLUNTARI_MOUSER}
       - subject: "RFQ Mouser - Sol·licitud de cotització #RFQ-2025-001"
       - body: [RFQ Mouser del context]
    
    3. Usa 'Send Email via Gmail' amb:
       - to_email: {VOLUNTARI_FARNELL}
       - subject: "RFQ Farnell - Sol·licitud de cotització #RFQ-2025-001"
       - body: [RFQ Farnell del context]
    """,
    expected_output=f"""
    ✅ Email 1/3 enviat a {VOLUNTARI_DIGIKEY}
    ✅ Email 2/3 enviat a {VOLUNTARI_MOUSER}
    ✅ Email 3/3 enviat a {VOLUNTARI_FARNELL}
    """,
    agent=agent_email_sender,
    context=[task_generate_rfqs]
)

### 6.6 Executar: Enviar RFQ

Ara executem el sistema complet fins aquest punt:

In [ ]:
crew_send_rfqs = Crew(
    agents=[agent_bom_analyzer, agent_rfq_generator, agent_email_sender],
    tasks=[
        task_analyze_bom,
        task_generate_rfqs,
        task_send_all_rfqs  # ← Una sola task que envia els 3
    ],
    process=Process.sequential,
    verbose=True
    max_iter=5,      # ← AFEGIR: Màxim 5 iteracions per tasca
    max_rpm=60       # ← AFEGIR: Màxim 60 requests/minut a l'API

)

print("🚀 Executant sistema complet: Analitzar BOM → Generar RFQs → Enviar als 3 Voluntaris\n")
print("=" * 80)
print(f"\n📧 S'enviaran RFQs a:")
print(f"   • DigiKey:  {VOLUNTARI_DIGIKEY}")
print(f"   • Mouser:   {VOLUNTARI_MOUSER}")
print(f"   • Farnell:  {VOLUNTARI_FARNELL}")
print("\n⏳ Això pot trigar 2-3 minuts...\n")
print("=" * 80)

result_send = crew_send_rfqs.kickoff()

print("\n" + "=" * 80)
print("✅ RFQs enviats correctament!")
print("=" * 80)
print("\n📊 RESULTAT:\n")
print(result_send)
print("\n" + "=" * 80)
print(f"\n💡 Els 3 voluntaris haurien de rebre els seus emails ara.")
print(f"   {VOLUNTARI_DIGIKEY} → RFQ DigiKey")
print(f"   {VOLUNTARI_MOUSER} → RFQ Mouser")
print(f"   {VOLUNTARI_FARNELL} → RFQ Farnell")

### 6.7 BREAK: Els 3 voluntaris responen (10-15 minuts)

**🕐 Ara farem un BREAK de 10-15 minuts perquè:**

**Els 3 voluntaris:**
1. ✅ Comprovin si han rebut el seu RFQ
2. ✅ Responguin simulant el seu proveïdor assignat:
   - **Voluntari 1** → Respon com a DigiKey (taula estructurada)
   - **Voluntari 2** → Respon com a Mouser (text lliure)
   - **Voluntari 3** → Respon com a Farnell (format caòtic)
3. ✅ Usin el template corresponent com a guia
4. ✅ Personalitzin lleugerament (preus, stock, etc.)

**La resta d'assistents:**
- 👀 Observen com funciona el procés
- ☕ Aprofiten per un cafè
- 🤔 Pensen com aplicar-ho als seus projectes

**🎯 Objectiu:** Tenir 3 respostes reals per processar amb el sistema!
**📋 Instruccions per respondre:**

1. Obre el teu Gmail
2. Busca l'email amb assumpte "Sol·licitud de cotització #RFQ-2025-001"
3. Respon usant el template del teu grup:
   - **Grup DigiKey:** Format taula estructurada
   - **Grup Mouser:** Text lliure conversacional
   - **Grup Farnell:** Format caòtic realista
4. **Inventa preus realistes** pels components (no cal que siguin reals)
5. Envia la resposta!

**💡 Consells:**
- Posa preus variats (alguns cars, alguns econòmics)
- Varia stocks (alguns "en stock", alguns "3-4 setmanes")
- Fes-ho divertit! Simula el teu proveïdor

In [ ]:
MY_SUPPLIER_ROLE = "Mouser" # Pot tenir valor "DigiKey" o "Mouser" o "Farnell"

# Mostra el template corresponent segons el rol
print(f"📋 EL TEU TEMPLATE ({MY_SUPPLIER_ROLE}):\n")
print("=" * 80)

if MY_SUPPLIER_ROLE == "DigiKey":
    print(TEMPLATE_DIGIKEY)
elif MY_SUPPLIER_ROLE == "Mouser":
    print(TEMPLATE_MOUSER)
elif MY_SUPPLIER_ROLE == "Farnell":
    print(TEMPLATE_FARNELL)
else:
    print("⚠️ Rol no reconegut! Tria: DigiKey, Mouser o Farnell")

print("\n" + "=" * 80)
print("\n💡 Copia aquest template, omple'l amb preus i respon l'email!")

---
## 7. Agent 3: Email Parser ⭐

**Aquesta és la part més interessant!** L'agent que parseja emails reals amb formats diversos.

### 7.1 Per què aquesta és la part clau?

**El desafiament:**
- ❌ Els emails NO estan estructurats
- ❌ Cada proveïdor usa un format diferent
- ❌ Pot faltar informació
- ❌ Poden haver errors tipogràfics
- ❌ Pot estar en idiomes diversos

**El poder dels LLMs:**
- ✅ Entenen context i significat
- ✅ Poden extreure info de text no estructurat
- ✅ Gestionen formats diversos
- ✅ Normalitzen dades automàticament
- ✅ Toleren errors i inconsistències

**Això NO es pot fer amb regex o parsers tradicionals!**

### 7.2 Crear l'Agent Email Parser

In [ ]:
agent_email_parser = Agent(
    role="Expert en Parsing d'Emails de Cotitzacions",
    goal="Extreure preus, stock i condicions d'emails de proveïdors amb formats diversos",
    backstory="""
    Ets un especialista en processament de llenguatge natural amb 10 anys d'experiència
    processant emails comercials.
    
    La teva expertesa és:
    - Llegir emails de proveïdors en formats MOLT diversos
    - Extreure informació clau: preus, stock, lead times, condicions
    - Normalitzar dades (convertir monedes, unitats, dates)
    - Gestionar informació incompleta o ambigua
    - Detectar warnings o condicions especials
    
    Pots processar:
    - Taules ben formatades
    - Text lliure conversacional
    - Formats caòtics o inconsistents
    - Múltiples idiomes (català, castellà, anglès)
    
    Tens accés a eines per llegir emails de Gmail.
    
    MOLT IMPORTANT:
    - Si no trobes una dada, marca-la com "No especificat" (NO inventes!)
    - Si hi ha ambigüitat, indica-ho a les notes
    - Converteix sempre a EUR si cal
    - Normalitza lead times a format estàndard (ex: "3-5 dies", "2 setmanes")
    
    El teu output ha de ser structured amb Pydantic per poder-lo processar després.
    """,
    tools=[fetch_emails_gmail],
    llm=LLM_MODEL
)

print("✅ Agent 'Email Parser' creat!")

### 7.3 Task: Parse Email Response

In [ ]:
task_parse_emails = Task(
    description=f"""
    Llegeix les 3 respostes als RFQs que hem enviat als voluntaris.
    
    Passos:
    1. Usa la tool 'Fetch Emails from Gmail' per llegir emails recents
       Cerca: "subject:cotització OR subject:RFQ newer_than:1d"
    
    2. Identifica les 3 respostes dels voluntaris:
       - DigiKey: {VOLUNTARI_DIGIKEY}
       - Mouser: {VOLUNTARI_MOUSER}
       - Farnell: {VOLUNTARI_FARNELL}
    
    3. Per CADA email rebut, extreu:
       - Nom del proveïdor (DigiKey/Mouser/Farnell)
       - Per CADA component mencionat:
         * Part Number (exacte)
         * Preu unitari (convertir a EUR si cal)
         * Stock disponible
         * Lead time
         * MOQ si està especificat
         * Notes especials
       - Informació general:
         * Cost d'enviament
         * Condicions de pagament
         * Validesa de la cotització
    
    CRÍTIC:
    - Gestiona formats diversos (taula, text lliure, caòtic)
    - Si una dada no està, marca-la com "No especificat" (NO inventes!)
    - Normalitza preus (sempre en EUR)
    - Processa les 3 respostes, no només una
    """,
    expected_output="""
    Dades extretes de les 3 respostes en format estructurat:
    
    ## Resposta 1: DigiKey
    - Email: {VOLUNTARI_DIGIKEY}
    - Data: [data]
    
    ### Components:
    - Part Number: [codi], Preu: €X.XX, Stock: [info], Lead time: [temps]
    [... més components ...]
    
    ### Condicions:
    - Enviament: €XX
    - Pagament: [terminis]
    - Validesa: [data]
    
    ## Resposta 2: Mouser
    [... mateix format ...]
    
    ## Resposta 3: Farnell
    [... mateix format ...]
    """,
    agent=agent_email_parser,
    tools=[fetch_emails_gmail]
)

print("✅ Task 'Parse Emails' creada per processar les 3 respostes!")

### 7.4 Executar Email Parser

**⚠️ Assegura't que:**
1. Han arribat els correus dels voluntaris.

In [ ]:
# Crew temporal per provar el parser
crew_parse = Crew(
    agents=[agent_email_parser],
    tasks=[task_parse_emails],  # ← Canviat de task_parse_email a task_parse_emails
    process=Process.sequential,
    verbose=False  # ← Desactivar verbose per evitar RecursionError
    max_iter=5,      # ← AFEGIR: Màxim 5 iteracions per tasca
    max_rpm=60       # ← AFEGIR: Màxim 60 requests/minut a l'API

)

print("🚀 Executant Email Parser...\n")
print("=" * 70)
print(f"\n📬 Buscant respostes de:")
print(f"   • DigiKey:  {VOLUNTARI_DIGIKEY}")
print(f"   • Mouser:   {VOLUNTARI_MOUSER}")
print(f"   • Farnell:  {VOLUNTARI_FARNELL}")
print("\n⏳ Llegint i processant emails...\n")
print("=" * 70)

result_parse = crew_parse.kickoff()

print("\n" + "=" * 70)
print("✅ Parsing completat!")
print("=" * 70)

# Verificar si és Pydantic
if hasattr(result_parse, 'pydantic') and result_parse.pydantic:
    parsed_data = result_parse.pydantic
    
    print("\n📋 DADES ESTRUCTURADES (Pydantic):\n")
    print(f"Proveïdor: {parsed_data.supplier_name}")  # ← Corregit
    print(f"Email: {parsed_data.supplier_email}")
    print(f"Data: {parsed_data.response_date}")
    
    print(f"\nComponents cotitzats: {len(parsed_data.components)}")  # ← Corregit
    for component in parsed_data.components:  # ← Corregit
        print(f"\n  • {component.part_number}")
        if component.description:
            print(f"    Descripció: {component.description}")
        print(f"    Preu: €{component.unit_price:.2f}")
        print(f"    Stock: {component.stock}")
        print(f"    Lead time: {component.lead_time}")
        if component.moq:
            print(f"    MOQ: {component.moq}")
        if component.notes:
            print(f"    Notes: {component.notes}")
    
    print(f"\n📦 Condicions:")
    print(f"  • Enviament: {parsed_data.shipping_cost}")
    print(f"  • Pagament: {parsed_data.payment_terms}")
    print(f"  • Validesa: {parsed_data.quote_validity}")
    if parsed_data.additional_notes:
        print(f"  • Notes: {parsed_data.additional_notes}")
else:
    # Si no és Pydantic, mostrar text pla
    print("\n📊 DADES EXTRETES:\n")
    print(result_parse)

### 💡 Observa la màgia!

**El que ha passat:**
1. ✅ L'agent ha llegit un email REAL
2. ✅ Ha entès el format (taula/text/caòtic)
3. ✅ Ha extret preus, stocks, lead times
4. ✅ Ha normalitzat les dades
5. ✅ Ha generat un objecte Pydantic estructurat

**Sense LLMs això seria:**
- ❌ Regex complexos per cada format
- ❌ Múltiples parsers específics
- ❌ Gestió manual d'excepcions
- ❌ Manteniment constant quan formats canvien

**Amb LLMs:**
- ✅ Un sol agent
- ✅ Gestiona tots els formats
- ✅ S'adapta automàticament
- ✅ Entén context i significat

---
## 8. Agents 4 i 5: Comparació i Informe

Ara que tenim les dades extretes, fem la comparativa i generem l'informe final.

### 8.1 Agent Price Comparator

In [ ]:
agent_price_comparator = Agent(
    role="Expert en Comparació de Preus i Condicions",
    goal="Comparar TOTES les cotitzacions dels 3 proveïdors i trobar la millor combinació global",
    backstory="""
    Ets un expert en procurement amb 20 anys d'experiència comparant ofertes de múltiples proveïdors.
    
    REGLES CRÍTIQUES:
    - Sempre analitzes TOTS els components de TOTS els proveïdors
    - NO descartes components que només té 1 proveïdor (són igualment importants!)
    - Crees una matriu completa: [TOTS els Components] x [3 Proveïdors]
    - Per cada component, mostres les 3 ofertes (o "No cotitzat" si no està)
    - Recomanes la millor combinació global que cobreix TOTS els components
    
    ESPECIALITAT:
    - Detectar oportunitats d'estalvi
    - Identificar riscos de proveïdor únic
    - Optimitzar cost total mantenint qualitat i disponibilitat
    - Generar recomanacions estratègiques de compra
    
    El teu objectiu és proporcionar una visió COMPLETA de totes les opcions disponibles.
    """,
    agent=agent_email_parser,
    max_iter=10  # Augmentat per processar més components
)

print("✅ Agent 'Price Comparator' creat amb enfocament COMPLET!")

### 8.2 Task: Compare Prices

### ✅ Verificació: Comptar Components de les 3 Cotitzacions

Abans d'executar la comparació, verifiquem que tenim TOTS els components:

In [ ]:
# Verificar que hem processat les 3 cotitzacions
import json

# El resultat del parser hauria d'estar disponible
# Simulem verificació (en producció vindria del context)

print("📊 Verificació de Components")
print("=" * 70)

# Comptadors esperats (segons els emails rebuts)
esperats = {
    "DigiKey": ["3005688", "3000836", "3003413", "3003412", "3003522", "3003494"],
    "Mouser": ["3003780", "3003681", "3003775", "3003762", "3003807", "3003808", "3003737", "3003669"],
    "Farnell": ["3003026", "3003060", "3003413", "3003412", "3003522", "3003494", "3003506"]
}

# Components únics totals
all_components = set()
for provider, components in esperats.items():
    all_components.update(components)

print(f"\n✅ Components per proveïdor:")
for provider, components in esperats.items():
    print(f"   • {provider:10s}: {len(components):2d} components")

print(f"\n✅ Components únics totals: {len(all_components)}")

# Components comuns
digikey_set = set(esperats["DigiKey"])
mouser_set = set(esperats["Mouser"])
farnell_set = set(esperats["Farnell"])

comuns_digikey_farnell = digikey_set & farnell_set
print(f"\n✅ Components comuns DigiKey-Farnell: {len(comuns_digikey_farnell)}")
print(f"   {sorted(comuns_digikey_farnell)}")

print(f"\n⚠️  L'agent comparador ha de processar TOTS els {len(all_components)} components!")
print(f"   NO només els {len(comuns_digikey_farnell)} comuns!")

print("\n" + "=" * 70)
print("💡 Si l'informe final té menys de {len(all_components)} components, hi ha un problema!")

---
### ⚠️ PROBLEMA COMÚ: Agent Comparador Incomplet

**Què pot passar:**
- L'agent només compara components que **coincideixen** entre proveïdors
- Ignora components que **només té 1 proveïdor**
- L'informe surt incomplet (només 6 components en lloc de 21)

**Per què passa:**
- Els LLMs tendeixen a comparar només quan hi ha **múltiples opcions**
- Sense instruccions explícites, ignoren components "no comparables"

**Solució aplicada:**
- ✅ Instruccions explícites: "NO excloguis cap component"
- ✅ Requeriment: "Inclou TOTS els part numbers de les 3 cotitzacions"
- ✅ Verificació: Comptador de components totals

**Exemple de l'error:**
```
Rebut:
- DigiKey: 6 components
- Mouser: 8 components  
- Farnell: 7 components
Total únic: 21 components

Informe erroni (sense solució):
- Només 6 components (els comuns DigiKey-Farnell)
- Mouser ignorat completament ❌

Informe correcte (amb solució):
- 21 components complets ✅
- Mostra "No cotitzat" quan no està disponible
- Comparativa completa
```

**Verificació:**
Al final de la comparació, comprova que el número de components coincideix amb el total únic de les 3 cotitzacions!

---

In [ ]:
task_compare_prices = Task(
    description="""
    Compara TOTES les cotitzacions dels 3 proveïdors (DigiKey, Mouser, Farnell).
    
    IMPORTANT: Has rebut cotitzacions de TOTS ELS COMPONENTS del BOM, distribuïts entre els 3 proveïdors.
    
    INSTRUCCIONS:
    
    1. Crea una llista COMPLETA de TOTS els components únics (Part Numbers) de les 3 cotitzacions
    
    2. Per CADA component de la llista:
       a) Verifica quins proveïdors l'han cotitzat:
          - DigiKey: té aquest component? → preu, stock, lead time
          - Mouser: té aquest component? → preu, stock, lead time  
          - Farnell: té aquest component? → preu, stock, lead time
       
       b) Si un proveïdor NO l'ha cotitzat, marca'l com "No cotitzat"
       
       c) Si 2+ proveïdors l'han cotitzat:
          - Compara preus
          - Compara stock
          - Compara lead times
          - Determina el millor
       
       d) Si només 1 proveïdor l'ha cotitzat:
          - Recomana aquest proveïdor (no hi ha alternativa)
    
    3. NO excloguis cap component! 
       - Components únics de DigiKey → inclou'ls
       - Components únics de Mouser → inclou'ls
       - Components únics de Farnell → inclou'ls
       - Components comuns → compara'ls
    
    4. Calcula costos totals:
       - Si compréssim TOT a DigiKey: €XXX (només components que té)
       - Si compréssim TOT a Mouser: €XXX (només components que té)
       - Si compréssim TOT a Farnell: €XXX (només components que té)
       - Millor combinació (components de diferents proveïdors): €XXX
    
    5. Detecta alertes:
       - Components amb stock <10 unitats
       - Lead times >3 setmanes
       - Components sense alternativa (només 1 proveïdor)
    
    CRÍTIC:
    - L'output ha d'incloure TOTS els part numbers de les 3 cotitzacions
    - Cada component ha de mostrar les 3 ofertes (o "No cotitzat")
    - Total de components ha de coincidir amb la suma de components únics de les 3 cotitzacions
    """,
    expected_output="""
    Informe de comparativa COMPLETA amb TOTS els components:
    
    ## Resum Executiu
    - Total components analitzats: [número total únic]
    - Components DigiKey: [número]
    - Components Mouser: [número]
    - Components Farnell: [número]
    - Components comuns: [número]
    - Cost òptim: €XXX.XX
    - Estratègia recomanada: [descripció]
    
    ## Anàlisi Detallada per Component
    
    ### [Part Number 1]
    - DigiKey: €X.XX/ud | Stock: XXX | Lead: Xd | [o "No cotitzat"]
    - Mouser: €X.XX/ud | Stock: XXX | Lead: Xd | [o "No cotitzat"]
    - Farnell: €X.XX/ud | Stock: XXX | Lead: Xd | [o "No cotitzat"]
    - RECOMANAT: [proveïdor] per [raó]
    
    [... TOTS els components de les 3 cotitzacions ...]
    
    ## Costos Totals per Estratègia
    
    | Estratègia | Components Disponibles | Cost Total | Observacions |
    |------------|----------------------|------------|--------------|
    | Tot DigiKey | X components | €XXX | Falten Y components |
    | Tot Mouser | X components | €XXX | Falten Y components |
    | Tot Farnell | X components | €XXX | Falten Y components |
    | Combinació Òptima | TOTS (X) | €XXX | Millor preu global |
    
    ## Recomanacions Finals
    1. Estratègia de compra recomanada
    2. Components crítics (1 sol proveïdor)
    3. Possibles estalvis
    """,
    agent=agent_price_comparator,
    context=[task_parse_emails]  # Rep les 3 cotitzacions parsejades
)

print("✅ Task 'Compare Prices' actualitzada amb anàlisi COMPLETA!")

### 8.3 Agent Report Generator

In [ ]:
agent_report_generator = Agent(
    role="Especialista en Informes Executius",
    goal="Generar informes professionals i clars per la direcció",
    backstory="""
    Ets un expert en comunicació corporativa amb 10 anys d'experiència.
    
    La teva especialitat és:
    - Generar informes executius clars i concisos
    - Presentar dades complexes de manera entenedora
    - Destacar insights clau (no només dades)
    - Usar visualitzacions quan cal
    - Adaptar el to segons l'audiència
    
    Per informes a direcció:
    - ✅ Comença amb el més important (resum executiu)
    - ✅ Usa números i percentatges (ROI, estalvis)
    - ✅ Destaca riscos i oportunitats
    - ✅ Proporciona recomanacions accionables
    - ✅ Format professional
    
    Els teus informes sempre responen:
    - Què hem trobat?
    - Quant ens costa?
    - Què recomanem?
    - Quins riscos hi ha?
    """,
    tools=[],
    llm=LLM_MODEL
)

print("✅ Agent 'Report Generator' creat!")

### 8.4 Task: Generate Final Report

In [ ]:
task_generate_report = Task(
    description="""
    Genera un informe executiu professional amb tota la informació de cotització.
    
    Basant-te en l'anàlisi comparativa anterior (context):
    
    1. Resum Executiu (el més important al principi):
       - Cost total recomanat
       - Estalvi vs alternatives
       - Principal recomanació
       - Riscos crítics
    
    2. Detall de Cotització:
       - Taula comparativa de components
       - Millor proveïdor per cada component
       - Costos desglosats
    
    3. Anàlisi de Riscos:
       - Components amb stock baix
       - Lead times llargs
       - Dependències d'un sol proveïdor
    
    4. Recomanacions:
       - Ordre de compra recomanat
       - Alternatives per components crítics
       - Accions de seguiment
    
    FORMAT: Markdown professional amb taules i seccions clares.
    TO: Executiu però tècnic (per R+D)
    LONGITUD: Complet però concís (2-3 pàgines)
    """,
    expected_output="""
    Un informe complet en format Markdown amb:
    
    # Informe de Cotització - [Projecte]
    *Data: [data]*
    *Generat per: Sistema automatitzat CrewAI*
    
    ## 📊 Resum Executiu
    [Paràgraf amb el més important]
    
    **Mètriques Clau:**
    - Cost total: €XXX.XX
    - Estalvi vs alternativa: €XX.XX (XX%)
    - Components cotitzats: XX
    - Proveïdors analitzats: X
    
    ## 💰 Detall de Cotització
    [Taula comparativa]
    
    ## ⚠️ Anàlisi de Riscos
    [Llista de riscos]
    
    ## 💡 Recomanacions
    [Llista de recomanacions accionables]
    
    ## 📝 Accions de Seguiment
    [Què cal fer ara]
    """,
    agent=agent_report_generator,
    context=[task_compare_prices]
)

print("✅ Task 'Generate Report' creada!")

---
## 9. Integració Completa del Sistema

Ara executem TOT el sistema de cap a cap!

### 9.1 Crear el Crew complet amb tots els agents

In [ ]:
from crewai import Agent, Task, Crew, Process

# Crew complet amb tots els agents i tasks
crew_complete_system = Crew(
    agents=[
        agent_bom_analyzer,
        agent_rfq_generator,
        agent_email_sender,
        # (Break: Companys responen)
        agent_email_parser,
        agent_price_comparator,
        agent_report_generator
    ],
    tasks=[
        task_analyze_bom,          # 1. Analitzar BOM
        task_generate_rfqs,        # 2. Generar RFQs
        task_send_all_rfqs,             # 3. Enviar per email
        # [BREAK aquí per respostes]
        task_parse_emails,          # 4. Parsear resposta
        task_compare_prices,       # 5. Comparar preus
        task_generate_report       # 6. Generar informe
    ],
    process=Process.sequential
)

print("✅ Crew complet creat amb 6 agents i 6 tasks!")
print("\n📊 Workflow:")
print("  1. BOM Analyzer → Identifica components")
print("  2. RFQ Generator → Genera emails")
print("  3. Email Sender → Envia via Gmail")
print("  4. [BREAK] → Companys responen")
print("  5. Email Parser → Extreu dades")
print("  6. Price Comparator → Analitza ofertes")
print("  7. Report Generator → Informe final")

### 9.2 Opció A: Executar tot (incloent l'enviament)

**⚠️ NOTA:** Això enviarà un email REAL. Només executa si vols fer tot el procés complet.

In [ ]:
# DESCOMENTAR per executar el sistema complet:

# print("🚀 EXECUTANT SISTEMA COMPLET DE COTITZACIÓ\n")
# print("=" * 70)
# print("\n⚠️ Aquest procés:")
# print("  1. Analitzarà el BOM")
# print("  2. Generarà RFQs")
# print(f"  3. Enviarà email a {PARTNER_EMAIL}")
# print("  4. Esperarà que responguis")
# print("  5. Parsearà la resposta")
# print("  6. Compararà preus")
# print("  7. Generarà informe final")
# print("\n⏳ Trigarà ~5 minuts (sense comptar el temps de resposta)\n")
# print("=" * 70)

# result_complete = crew_complete_system.kickoff()

# print("\n" + "=" * 70)
# print("✅ SISTEMA COMPLET EXECUTAT!")
# print("=" * 70)
# print("\n📊 INFORME FINAL:\n")
# print(result_complete)

print("💡 Descomenta el codi de dalt per executar el sistema complet.")
print("   O continua amb l'Opció B per executar només des del parsing.")

### 9.3 Opció B: Executar només parsing + comparació + informe

Si ja has enviat l'RFQ manualment i tens respostes, usa això:

In [ ]:
from crewai import Agent, Task, Crew, Process

# ═══════════════════════════════════════════════════════════
# CREW: Parsing → Comparació → Informe
# ═══════════════════════════════════════════════════════════

crew_parse_to_report = Crew(
    agents=[agent_email_parser, agent_price_comparator, agent_report_generator],
    tasks=[task_parse_emails, task_compare_prices, task_generate_report],
    process=Process.sequential,
    verbose=True  
)

print("🚀 Executant sistema complet: Parsing → Comparació → Informe\n")
print("=" * 80)
print("\n📬 Buscant respostes dels 3 voluntaris:")
print(f"   • DigiKey:  {VOLUNTARI_DIGIKEY}")
print(f"   • Mouser:   {VOLUNTARI_MOUSER}")
print(f"   • Farnell:  {VOLUNTARI_FARNELL}")
print("\n⏳ Processant les 3 respostes...\n")
print("   1️⃣  Parsejar emails rebuts")
print("   2️⃣  Comparar preus entre proveïdors")
print("   3️⃣  Generar informe executiu")
print("\n" + "=" * 80)

# Executar
result_parse_to_report = crew_parse_to_report.kickoff()

print("\n" + "=" * 80)
print("✅ PROCÉS COMPLETAT!")
print("=" * 80)
print("\n📊 INFORME FINAL:\n")
print(result_parse_to_report)
print("\n" + "=" * 80)

### 9.4 Guardar l'informe

Guardem l'informe generat a un fitxer Markdown:

In [ ]:
# ═══════════════════════════════════════════════════════════
# GUARDAR INFORME
# ═══════════════════════════════════════════════════════════

import os
from datetime import datetime
from IPython.display import Markdown, display

# Crear directori d'outputs si no existeix
os.makedirs('outputs', exist_ok=True)

# Nom del fitxer amb timestamp
timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
filename = f"outputs/informe_cotitzacio_{timestamp}.md"

# Guardar informe
with open(filename, 'w', encoding='utf-8') as f:
    f.write(str(result_parse_to_report))

print(f"\n💾 Informe guardat a: {filename}")

# Mostrar informe formatat al notebook
print("\n" + "=" * 80)
print("📄 INFORME FORMATAT")
print("=" * 80)
print()

display(Markdown(str(result_parse_to_report)))

print("\n💡 L'informe també s'ha guardat com a fitxer Markdown.")

---
## 🎉 Felicitats!

Has completat el cas pràctic complet i has construït un sistema real i funcional!

**El que has après:**
- ✅ Crear sistemes multi-agent complexos
- ✅ Integrar amb Gmail API (email real)
- ✅ Parsing intel·ligent de text no estructurat
- ✅ Structured outputs amb Pydantic
- ✅ Workflows seqüencials amb dependències
- ✅ Gestió d'errors i fallbacks
- ✅ Generació d'informes professionals

**El que has construït:**
Un sistema que:
1. Analitza BOMs automàticament
2. Genera i envia RFQs per email
3. Parseja respostes en formats diversos
4. Compara preus i condicions
5. Genera informes executius

**I tot això amb només 6 agents i ~300 línies de codi! 🚀**

**Properes passes:**
- Adapta'l al teu cas d'ús específic a JCM
- Integra amb els vostres sistemes (ERP, Outlook, etc.)
- Afegeix més agents per funcionalitats addicionals
- Comparteix amb l'equip i itereu junts

**Gràcies per participar en aquesta formació! 🙏**